### PR TITLE
Handle KAE (or other exceptions of similar type) in the future when witnessing transactions

### PR DESCRIPTION
**Pull request by Sam-Kramer**
## General
**Before this PR**:

**After this PR**:

==COMMIT_MSG==
Handles transactions that may or may have not been witnessed due to errors (such as KeyAlreadyExistsException) in the workload server. 
==COMMIT_MSG==

**Priority**:

**Concerns / possible downsides (what feedback would you like?)**:

**Is documentation needed?**:

## Compatibility
**Does this PR create any API breaks (e.g. at the Java or HTTP layers) - if so, do we have compatibility?**:

**Does this PR change the persisted format of any data - if so, do we have forward and backward compatibility?**:

**The code in this PR may be part of a blue-green deploy. Can upgrades from previous versions safely coexist? (Consider restarts of blue or green nodes.)**:

**Does this PR rely on statements being true about other products at a deployment - if so, do we have correct product dependencies on these products (or other ways of verifying that these statements are true)?**:

**Does this PR need a schema migration?**

## Testing and Correctness
**What, if any, assumptions are made about the current state of the world? If they change over time, how will we find out?**:

**What was existing testing like? What have you done to improve it?**:

**If this PR contains complex concurrent or asynchronous code, is it correct? The onus is on the PR writer to demonstrate this.**:

**If this PR involves acquiring locks or other shared resources, how do we ensure that these are always released?**:

## Execution
**How would I tell this PR works in production? (Metrics, logs, etc.)**:

**Has the safety of all log arguments been decided correctly?**:

**Will this change significantly affect our spending on metrics or logs?**:

**How would I tell that this PR does not work in production? (monitors, etc.)**:

**If this PR does not work as expected, how do I fix that state? Would rollback be straightforward?**:

**If the above plan is more complex than “recall and rollback”, please tag the support PoC here (if it is the end of the week, tag both the current and next PoC)**:

## Scale
**Would this PR be expected to pose a risk at scale? Think of the shopping product at our largest stack.**:

**Would this PR be expected to perform a large number of database calls, and/or expensive database calls (e.g., row range scans, concurrent CAS)?**:

**Would this PR ever, with time and scale, become the wrong thing to do - and if so, how would we know that we need to do something differently?**:

## Development Process
**Where should we start reviewing?**:

**If this PR is in excess of 500 lines excluding versions lock-files, why does it not make sense to split it?**:

**Please tag any other people who should be aware of this PR**:
@jeremyk-91
@sverma30
@raiju

### CODE REVIEW

File: atlasdb-workload-server/src/main/java/com/palantir/atlasdb/workload/workflow/DefaultWorkflow.java [hunk 1]
```diff
@@ -60,7 +61,7 @@ public static <T extends TransactionStore> Workflow create(
     @Override
     public WorkflowHistory run() {
```
Comment (Sam-Kramer) [inline atlasdb-workload-server/src/main/java/com/palantir/atlasdb/workload/workflow/DefaultWorkflow.java:62]: We may wish to update the `WorkflowHistory` class that it only accepts `FullyWitnessedTransaction` rather than both classes. I think this is ok for now.

File: atlasdb-workload-server/src/test/java/com/palantir/atlasdb/workload/store/AtlasDbTransactionStoreTest.java [hunk 2]
```diff
@@ -162,6 +166,29 @@ public void abortedTransactionsReturnEmpty() {
...
+                    throw new KeyAlreadyExistsException("foobar");
+                })
+                .when(onlyThrowsKeyAlreadyExistsExceptionManager)
+                .runTaskWithConditionWithRetry(any(Supplier.class), any());
+        AtlasDbTransactionStore onlyKeyAlreadyExistsThrowingStore =
+                AtlasDbTransactionStore.create(onlyThrowsKeyAlreadyExistsExceptionManager, TABLES_TO_ATLAS_METADATA);
+        assertThat(onlyKeyAlreadyExistsThrowingStore.readWrite(List.of(writeTransactionAction)))
+                .hasValueSatisfying(witnessedTransaction -> {
+                    assertThat(witnessedTransaction).isInstanceOf(MaybeWitnessedTransaction.class);
+                    assertThat(witnessedTransaction.actions()).containsExactly(writeTransactionAction.witness());
```
Comment (Sam-Kramer) [inline atlasdb-workload-server/src/test/java/com/palantir/atlasdb/workload/store/AtlasDbTransactionStoreTest.java:188]: re-reading this, probably should have checked that commit timestamp is present, but this is a precondition when we construct `MaybeWitnessedTransaction`, thus is not needed

File: atlasdb-workload-server-api/src/main/java/com/palantir/atlasdb/workload/transaction/witnessed/MaybeWitnessedTransaction.java [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+ * unsure if they have committed or not due to failures.
+ */
+@Value.Immutable
+public interface MaybeWitnessedTransaction extends WitnessedTransaction {
+    @Value.Check
+    default void check() {
+        Preconditions.checkArgument(
+                commitTimestamp().isPresent(),
+                "Commit timestamp must be present in a potentially witnessed transaction, as otherwise it is not"
+                        + " possible to validate whether or not a transaction has been committed.");
```
Comment (jeremyk-91) [inline atlasdb-workload-server-api/src/main/java/com/palantir/atlasdb/workload/transaction/witnessed/MaybeWitnessedTransaction.java:33]: I think this can be stronger: something like `given how the transaction protocol works, a transaction for which we haven't retrieved the commit timestamp has not written anything of significance to the database` or something like that?

File: atlasdb-workload-server-api/src/main/java/com/palantir/atlasdb/workload/store/ReadableTransactionStore.java [hunk 4]
```diff
@@ -28,4 +28,9 @@ public interface ReadableTransactionStore {
      * @return The value of the cell for a given table.
      */
     Optional<Integer> get(String table, WorkloadCell cell);
+
+    /**
+     * Checks the startTimestamp for a given transaction whether it has committed.
```
Comment (jeremyk-91) [inline atlasdb-workload-server-api/src/main/java/com/palantir/atlasdb/workload/store/ReadableTransactionStore.java:33]: nit: checks whether the transaction with the provided startTimestamp has committed?

File: atlasdb-workload-server/src/test/java/com/palantir/atlasdb/workload/transaction/witnessed/OnlyCommittedWitnessedTransactionVisitorTest.java [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+
+    @Mock
+    private ReadOnlyTransactionStore transactionStore;
+
+    @Test
+    public void fullyCommittedTransactionsAreNeverChecked() {
+        lenient().when(transactionStore.isCommitted(anyLong())).thenThrow(new RuntimeException());
+        assertThat(READ_ONLY_WITNESSED_TRANSACTION.accept(
+                        new OnlyCommittedWitnessedTransactionVisitor(transactionStore)))
+                .contains(READ_ONLY_WITNESSED_TRANSACTION);
```
Comment (jeremyk-91) [inline atlasdb-workload-server/src/test/java/com/palantir/atlasdb/workload/transaction/witnessed/OnlyCommittedWitnessedTransactionVisitorTest.java:49]: For readability, this kind of check is normally done by asserting `verifyNoInteractions(transactionStore);` (also, while it doesn't apply here, an implementation that catches exceptions and does stuff with them could pass this test as written!)

File: atlasdb-workload-server/src/main/java/com/palantir/atlasdb/workload/workflow/DefaultWorkflow.java (hunk 1 above)
Comment (jeremyk-91) [reply atlasdb-workload-server/src/main/java/com/palantir/atlasdb/workload/workflow/DefaultWorkflow.java:63]: Yep. Could we leave a todo here? `WorkflowHistory` should only accept `FullyWitnessedTransaction` long term, yeah.

File: atlasdb-workload-server-api/src/test/java/com/palantir/atlasdb/workload/transaction/witnessed/MaybeWitnessedTransactionTest.java [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+import java.util.List;
+import org.junit.Test;
+
+public class MaybeWitnessedTransactionTest {
+
+    private static final Long START_TIMESTAMP = 100L;
+    private static final Long COMMIT_TIMESTAMP = 200L;
+
+    @Test
+    public void throwsWhenCommitTimestampIsNotPresent() {
```
Comment (jeremyk-91) [inline atlasdb-workload-server-api/src/test/java/com/palantir/atlasdb/workload/transaction/witnessed/MaybeWitnessedTransactionTest.java:34]: nit: these two should mention `creation` or something like that

File: atlasdb-workload-server/src/main/java/com/palantir/atlasdb/workload/store/AtlasDbTransactionStore.java [hunk 7]
```diff
@@ -99,9 +113,22 @@ public Optional<WitnessedTransaction> readWrite(Consumer<InteractiveTransaction>
                     .build());
         } catch (SafeIllegalArgumentException e) {
             throw e;
+        } catch (KeyAlreadyExistsException e) {
+            Transaction transaction = transactionReference.get();
+            return Optional.of(MaybeWitnessedTransaction.builder()
+                    .startTimestamp(transaction.getTimestamp())
+                    .commitTimestamp(commitTimestampProvider
+                            .get()
+                            .getCommitTimestampOrThrowIfMaybeNotCommitted(transaction.getTimestamp()))
```
Comment (jeremyk-91) [inline atlasdb-workload-server/src/main/java/com/palantir/atlasdb/workload/store/AtlasDbTransactionStore.java:122]: not actionable: This might work strangely if the transaction task deliberately throws a `KeyAlreadyExistsException`, though I don't think we need to worry about that

File: atlasdb-workload-server/src/main/java/com/palantir/atlasdb/workload/store/AtlasDbTransactionStore.java [hunk 8]
```diff
@@ -99,9 +113,22 @@ public Optional<WitnessedTransaction> readWrite(Consumer<InteractiveTransaction>
...
+                    .build());
         } catch (Exception e) {
-            // TODO: Need to eventually handle PuE exceptions, as they could've succeeded in committing.
-            log.info("Failed to record transaction due to an exception", e);
+            Optional<Long> startTimestamp =
+                    Optional.ofNullable(transactionReference.get()).map(Transaction::getTimestamp);
+            log.info(
+                    "Failed to record transaction due to an exception for startTimestamp {}",
+                    SafeArg.of("startTimestamp", startTimestamp),
+                    e);
```
Comment (jeremyk-91) [inline atlasdb-workload-server/src/main/java/com/palantir/atlasdb/workload/store/AtlasDbTransactionStore.java:131]: We might want to set the transaction reference at the beginning of the task, if we're concerned about something going wrong before we enter the commit-writes phase.

File: atlasdb-workload-server/src/test/java/com/palantir/atlasdb/workload/store/AtlasDbTransactionStoreTest.java [hunk 9]
```diff
@@ -162,6 +166,29 @@ public void abortedTransactionsReturnEmpty() {
                 .isEmpty();
     }
 
+    @Test
+    public void keyAlreadyExistExceptionResultsInMaybeWitnessedTransaction() {
+        WriteTransactionAction writeTransactionAction =
+                WriteTransactionAction.of(TABLE_1, WORKLOAD_CELL_ONE, VALUE_ONE);
+        TransactionManager onlyThrowsKeyAlreadyExistsExceptionManager = spy(manager);
+        doAnswer(answer -> {
```
Comment (jeremyk-91) [inline atlasdb-workload-server/src/test/java/com/palantir/atlasdb/workload/store/AtlasDbTransactionStoreTest.java:174]: nit: `Answer` is a functional interface with a method that takes an `InvocationOnMock`, so this param should be named `invocation` or `invocationOnMock`

File: atlasdb-workload-server/src/test/java/com/palantir/atlasdb/workload/store/AtlasDbTransactionStoreTest.java [hunk 10]
```diff
@@ -162,6 +166,29 @@ public void abortedTransactionsReturnEmpty() {
...
+                    ConditionAwareTransactionTask<Void, CommitTimestampProvider, Exception> task =
+                            answer.getArgument(1);
+                    manager.runTaskWithConditionWithRetry(commitTimestampFetcher, task);
+                    throw new KeyAlreadyExistsException("foobar");
+                })
+                .when(onlyThrowsKeyAlreadyExistsExceptionManager)
+                .runTaskWithConditionWithRetry(any(Supplier.class), any());
+        AtlasDbTransactionStore onlyKeyAlreadyExistsThrowingStore =
+                AtlasDbTransactionStore.create(onlyThrowsKeyAlreadyExistsExceptionManager, TABLES_TO_ATLAS_METADATA);
+        assertThat(onlyKeyAlreadyExistsThrowingStore.readWrite(List.of(writeTransactionAction)))
```
Comment (jeremyk-91) [inline atlasdb-workload-server/src/test/java/com/palantir/atlasdb/workload/store/AtlasDbTransactionStoreTest.java:185]: Hmm, I think the naming here is a bit hard to follow - not sure about the significance of the `only`. Maybe `keyAlreadyExistsExceptionThrowingManager` and `keyAlreadyExistsExceptionThrowingStore`?

File: atlasdb-workload-server-api/src/main/java/com/palantir/atlasdb/workload/transaction/witnessed/MaybeWitnessedTransaction.java [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+    static ImmutableMaybeWitnessedTransaction.Builder builder() {
+        return ImmutableMaybeWitnessedTransaction.builder();
+    }
+
+    default FullyWitnessedTransaction toFullyWitnessed() {
+        return FullyWitnessedTransaction.builder()
+                .startTimestamp(startTimestamp())
+                .commitTimestamp(commitTimestamp())
+                .actions(actions())
+                .build();
```
Comment (jeremyk-91) [inline atlasdb-workload-server-api/src/main/java/com/palantir/atlasdb/workload/transaction/witnessed/MaybeWitnessedTransaction.java:50]: non actionable: If we add an optional field to `WitnessedTransaction` after this, we won't copy over its value from the `Maybe` to the `Fully` version. I think this risk is ok. Maybe there's some clever refactoring we can do, but not needed here.

File: atlasdb-workload-server/src/main/java/com/palantir/atlasdb/workload/store/AtlasDbTransactionStore.java (hunk 7 above)
Comment (Sam-Kramer) [reply atlasdb-workload-server/src/main/java/com/palantir/atlasdb/workload/store/AtlasDbTransactionStore.java:122]: Later on, I made it the more general case of `TransactionCommitFailedException`, rather than just KAE, but still, a transaction could also throw this for funs. I think in this case we're ok, as we validate later on if it exists or not

**Review by jeremyk-91 — APPROVED** [review]
👍
